### PR TITLE
Parallel polydata reader

### PR DESCRIPTION
**Pull request by Davknapp**
Implemented a parallel reader for vtkPolyData. 
Restructured the parallel reader to decrease Code-duplication. 

The merging of polyData and vtkUnstructuredGrids are slightly different, resulting in two slightly different merging-functions. 

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

#### General
- [x] The reviewer executed the new code features at least once and checked the results manually

- [x] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [x] New source/header files are properly added to the Makefiles
- [x] The code is well documented
- [x] All function declarations, structs/classes and their members have a proper doxygen documentation
- [x] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)

#### Tests
- [x] The code is covered in an existing or new test case using Google Test

#### Github action

- [x] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)
- [x] All tests pass (in various configurations, this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

#### Scripts and Wiki

- [x] If a new directory with source-files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [x] New Datatypes are added to t8indent_custom_datatypes.txt
- [x] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

#### Licence

- [x] The author added a BSD statement to `doc/` (or already has one)

### CODE REVIEW

**Comment by jmark** [comment]
@Davknapp Please fix the merge conflicts and update the branch if applicable. Then I start the review.

File: src/t8_vtk/t8_vtk_parallel.cxx [hunk 1]
```diff
@@ -63,37 +125,78 @@ t8_read_parallel (const char *filename, vtkSmartPointer < vtkDataSet > grid,
...
-    vtkNew < vtkAppendFilter > append;
-    for (int ipiece = first_piece; ipiece < last_piece; ipiece++) {
-      reader->UpdatePiece (ipiece, total_num_pieces, 0);
-      append->AddInputData (reader->GetOutput ());
-    }
-    /* Merge all read grids together */
-    append->Update ();
-    append->MergePointsOn ();
-    grid->ShallowCopy (append->GetOutput ());
+    merge_unstructured (reader, grid, first_piece, last_piece);
```
Comment (jmark) [inline src/t8_vtk/t8_vtk_parallel.cxx:199]: Is it necessary to have a routine `merge_unstructured` just for this one call?

File: src/t8_vtk/t8_vtk_parallel.hxx [hunk 2]
```diff
@@ -32,9 +32,9 @@ along with t8code; if not, write to the Free Software Foundation, Inc.,
...
 /**
  * Given a filename to a parallel vtk file (for example .pvtu) and its data files, 
- * read a piece of the data files (like .vtu, .vtp, ...).
+ * read a piece of the data files (like .vtu, ...).
  * 
- * \param[in] filename  The name of a parallel vtk file (.pvtu for example)
+ * \param[in] filename  The name of a parallel vtk file to a distributed vtkUnstructuredGrid
  * \param[in] grid On input a vtkSmartPointer, that will hold the grid described
  *                 by the pieces read on this proc. 
  * \returns        non-zero on success, zero if the reading failed.
```
Comment (jmark) [inline src/t8_vtk/t8_vtk_parallel.hxx:40]: I recommend to be explicit: `t8_read_parallel_unstructured`.

File: src/t8_vtk/t8_vtk_reader.cxx [hunk 3]
```diff
@@ -149,6 +151,16 @@ t8_file_to_vtkGrid (const char *filename,
         break;
       }
       break;
+    case VTK_PARALLEL_POLYDATA_FILE:
+      if (!partition) {
+        main_proc_read_successful = t8_read_poly (filename, vtkGrid);
```
Comment (jmark) [inline src/t8_vtk/t8_vtk_reader.cxx:156]:
```suggestion
        main_proc_read_successful = t8_read_polyData (filename, vtkGrid);
```

File: src/t8_vtk/t8_vtk_reader.cxx [hunk 4]
```diff
@@ -149,6 +151,16 @@ t8_file_to_vtkGrid (const char *filename,
...
+      }
+      else {
+        main_proc_read_successful =
+          t8_read_parallel_poly (filename, vtkGrid, comm);
```
Comment (jmark) [inline src/t8_vtk/t8_vtk_reader.cxx:160]:
```suggestion
          t8_read_parallel_polyData (filename, vtkGrid, comm);
```

File: test/t8_IO/t8_gtest_vtk_reader.cxx [hunk 5]
```diff
@@ -45,37 +45,40 @@ class vtk_reader : public testing::TestWithParam<std::tuple<int, int, int>>{
...
         GTEST_SKIP ();
       } 
       file = std::get<0>(GetParam());
       file_type = gtest_vtk_filetypes[file];
       partition = std::get<1>(GetParam());
       main_proc = std::get<2>(GetParam());
-      if((file_type & VTK_PARALLEL_POLYDATA_FILE )&& partition){
-        GTEST_SKIP();
-      }
+      distributed = (file_type & VTK_PARALLEL_FILE) && partition;
```
Comment (jmark) [inline test/t8_IO/t8_gtest_vtk_reader.cxx:57]: Isn't this a cool line of code? It reads like a sentence and states expliclitly what the author wants to express here. :)

**Review by jmark — CHANGES_REQUESTED** [review]
Would you also update `t8_cmesh_read_from_vtk` to be able to read paralle polydata files?

File: src/t8_vtk/t8_vtk_parallel.cxx [hunk 6]
```diff
@@ -63,37 +125,78 @@ t8_read_parallel (const char *filename, vtkSmartPointer < vtkDataSet > grid,
...
+  int                 first_piece = 0;
+  int                 last_piece = -1;
+  read_status =
+    setup_reader (filename, reader, &first_piece, &last_piece, comm);
+  if (read_status == read_failure) {
+    return read_status;
+  }
+  /* Read the pieces if there are any pieces to read on this proc. */
+  if (first_piece < last_piece) {
+    merge_polydata (reader, grid, first_piece, last_piece);
```
[the remark this answers is not in the document]
Comment (Davknapp) [reply src/t8_vtk/t8_vtk_parallel.cxx:170]: Would you prefer the function inside of the if-block?

File: test/t8_IO/t8_gtest_vtk_reader.cxx (hunk 5 above)
Comment (Davknapp) [reply test/t8_IO/t8_gtest_vtk_reader.cxx:57]: It is, indeed!

File: src/t8_vtk/t8_vtk_parallel.cxx (hunk 6 above)
[the remark this answers is not in the document]
Comment (jmark) [reply src/t8_vtk/t8_vtk_parallel.cxx:170]: Indeed. The function is not very big and fits well in the if block.

[the remark this answers is not in the document]
Comment (Davknapp) [reply src/t8_vtk/t8_vtk_parallel.cxx]: Done

File: src/t8_vtk/t8_vtk_parallel.hxx (hunk 2 above)
Comment (Davknapp) [reply src/t8_vtk/t8_vtk_parallel.hxx:40]: Done

**Review by jmark — CHANGES_REQUESTED** [review]
LGTM!

Just one final request.

Would you update `t8_cmesh_read_from_vtk` to be able to read paralle polydata files?

I had this comment in the last review, too. Maybe you overlooked it?

File: example/IO/cmesh/vtk/t8_cmesh_read_from_vtk.cxx [hunk 7]
```diff
@@ -166,9 +166,9 @@ main (int argc, char **argv)
                        "If set, partition the cmesh uniformly.");
   sc_options_add_int (opt, 't', "type_of_file", &vtk_file_type_int, -1,
                       "Set the type of the data in the file.\n"
-                      "\t\t\t\t\t0 for vtkUnstructuredGrid \n"
-                      "\t\t\t\t\t1 for vtkPolyData\n"
-                      "\t\t\t\t\t2 for pvtu.");
+                      "\t\t\t\t\t0 for vtkUnstructuredGrid,\n"
+                      "\t\t\t\t\t1 for vtkPolyData,\n"
+                      "\t\t\t\t\t2 for pvtu," "\t\t\t\t\t3 for pvtp,");
```
Comment (sandro-elsweijer) [inline example/IO/cmesh/vtk/t8_cmesh_read_from_vtk.cxx:171]:
```suggestion
                      "\t\t\t\t\t2 for pvtu,\n" "\t\t\t\t\t3 for pvtp.");
```

**Review by sandro-elsweijer — CHANGES_REQUESTED** [review]

File: src/t8_vtk/t8_vtk_polydata.hxx [hunk 8]
```diff
@@ -47,7 +47,8 @@ along with t8code; if not, write to the Free Software Foundation, Inc.,
  * \returns             non-zero on success, zero if the reading failed.              
  * 
  */
-vtk_read_success_t  t8_read_poly (const char *filename, vtkDataSet * grid);
+vtk_read_success_t  t8_read_polyData (const char *filename,
+                                      vtkDataSet * grid);
```
[the remark this answers is not in the document]
Comment (Davknapp) [reply src/t8_vtk/t8_vtk_polydata.hxx:51]: Oh, that is a good catch! Thank you, we should've done this earlier

[the remark this answers is not in the document]
Comment (Davknapp) [reply src/t8_vtk/t8_vtk_polydata.hxx:51]: Fixed it on a seperate branch, see PR #668

**Review by sandro-elsweijer — APPROVED** [review]

**Pull request merged**